diffrax/integrate.py

```diff
@@ -38,6 +40,48 @@
 from .term import AbstractTerm, VectorFieldWrapper, WrapTerm
 
 
+class _HistoryVectorField(eqx.Module):
+    vector_field: Callable
+    dense_interp: Optional[DenseInterpolation]
+    y0_history: Callable
+    delays: PyTree[Callable]
+    t0: float
+
+    def __call__(self, t, y, args):
+        history_vals = []
+        delays, treedef = jtu.tree_flatten(self.delays)
+        if self.dense_interp is None:
+            for delay in self.delays:
+                delay_val = delay(t, y, args)
+                alpha_val = t - delay_val
+                y0_val = self.y0_history(alpha_val)
+                history_vals.append(y0_val)
+        else:
+            for delay in delays:
+                delay_val = delay(t, y, args)
+                alpha_val = t - delay_val
+                history_delay_val = jnp.where(alpha_val > self.t0, alpha_val, self.t0)
+                history_val = self.dense_interp.evaluate(history_delay_val)
+                y0_delay_val = jnp.where(alpha_val > self.t0, self.t0, alpha_val)
+                y0_val = self.y0_history(y0_delay_val)
+                real_history_val = jnp.where(self.t0 > alpha_val, y0_val, history_val)
+                history_vals.append(real_history_val)
+
+        history_vals = jtu.tree_unflatten(treedef, history_vals)
+        history_vals = tuple(history_vals)
+        return self.vector_field(t, y, args, history=history_vals)
+
+
+class _Delays(eqx.Module):
+    delays: Optional[PyTree[Callable]]
```

Comment:
This should be public API, i.e. `Delays` rather than `_Delays`.

Also, I think `delays` should have type `PyTree[Callable]`. The optionality is captured via `Optional[Delays]` below.

Reply:
yes gotcha